File: Rapid React 6364/src/main/java/frc/robot/subsystems/DriveSubsystem.java
package frc.robot.subsystems;

//Import Subsystem Module
import edu.wpi.first.wpilibj2.command.SubsystemBase;

//Import Motor Modules
import com.ctre.phoenix.motorcontrol.ControlMode;
import com.ctre.phoenix.motorcontrol.NeutralMode;
import com.ctre.phoenix.motorcontrol.can.TalonFX;

//Import Gyro Modules
import com.kauailabs.navx.frc.AHRS;
import edu.wpi.first.wpilibj.SPI;

//WPI
import edu.wpi.first.math.controller.PIDController;
import edu.wpi.first.math.controller.SimpleMotorFeedforward;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.geometry.Pose2d;
import edu.wpi.first.math.kinematics.DifferentialDriveKinematics;
import edu.wpi.first.math.kinematics.DifferentialDriveOdometry;
import edu.wpi.first.math.kinematics.DifferentialDriveWheelSpeeds;
import edu.wpi.first.wpilibj.DriverStation;

//Import Robot Files
import frc.robot.utilities.InitializeTalon;
import frc.robot.Constants;

public class DriveSubsystem extends SubsystemBase{

    // Declare Drive Motors - driveRM = Drive Right Master // driveLS  = Drive Left Slave
    private TalonFX driveRM;
    private TalonFX driveRS;
    private TalonFX driveLM;
    private TalonFX driveLS;

    //Declare Sensors
    private AHRS gyro;

    //Declare Others
    private DifferentialDriveKinematics iKinematics;
    private DifferentialDriveWheelSpeeds iSpeeds;
    private DifferentialDriveOdometry iOdometry;
    private SimpleMotorFeedforward feedforward;
    private PIDController rightPID;
    private PIDController leftPID;

    public DriveSubsystem() {
        try {
            driveRM = new TalonFX(Constants.RobotMap.driveRM.port);
            driveRS = new TalonFX(Constants.RobotMap.driveRS.port);
            driveLM = new TalonFX(Constants.RobotMap.driveLM.port);
            driveLS = new TalonFX(Constants.RobotMap.driveLS.port);
        } catch (RuntimeException ex) {
            DriverStation.reportError("Error Starting Drive Motors: " + ex.getMessage(), true);
        }

        //Initialize Motors
        InitializeTalon.RightDrive(driveRM);
        InitializeTalon.RightDrive(driveRS);
        InitializeTalon.LeftDrive(driveLM);
        InitializeTalon.LeftDrive(driveLS);

        //Set Drive Brake Modes
        driveRM.setNeutralMode(NeutralMode.Coast);
        driveLM.setNeutralMode(NeutralMode.Coast);

        //Set Slave Motors To Follow Masters
        driveRS.follow(driveRM);
        driveLS.follow(driveLM);

        try {
            gyro = new AHRS(SPI.Port.kMXP);
        } catch (RuntimeException ex) {
            DriverStation.reportError("Error Starting NavX: " + ex.getMessage(), true);
        }
        gyro.reset();

        iOdometry = new DifferentialDriveOdometry(Rotation2d.fromDegrees(getAngle()));
        iKinematics = new DifferentialDriveKinematics(Constants.Drive.kTrackWidthMeters);
        iSpeeds = new DifferentialDriveWheelSpeeds(getLeftVelocity(), getRightVelocity());
        feedforward = new SimpleMotorFeedforward(Constants.Drive.ksVolts, Constants.Drive.kvVoltSecondsPerMeter, Constants.Drive.kaVoltSecondsSquaredPerMeter);
        rightPID = new PIDController(Constants.Drive.kPDriveVel, 0, 0);
        leftPID = new PIDController(Constants.Drive.kPDriveVel, 0, 0);
    }

    /**
     * Set Drive Motors Control Mode and Power
     * @param Mode CTRE TalonFX Control Mode
     * @param rightVal The Right Power Value
     * @param leftVal The Left Power Value
     */
    public void set(ControlMode Mode, double rightVal, double leftVal){
        driveRM.set(Mode, rightVal);
        driveLM.set(Mode, leftVal);
    }

    /**
     * Stops all Drivebase Movement
     */
    public void stop(){
        driveRM.set(ControlMode.PercentOutput, 0);
        driveLM.set(ControlMode.PercentOutput, 0);
    }

    /**
     * Get The Right Drive Position in Encoder Units (2408 per Revolute)
     * @return Current Position in Encoder Units
     */
    public double getRightPosition(){
        return driveRM.getSelectedSensorPosition();
    }

     /**
     * Get The Left Drive Position in Encoder Units (2408 per Revolute)
     * @return Current Position in Encoder Units
     */   
    public double getLeftPosition(){
        return driveLM.getSelectedSensorPosition();
    }

    /**
     * Get The Current Right Drive Velocity In Encoder Units
     * @return Encoder Units per 100ms
     */
    public double getRightVelocity(){
        return driveRM.getSelectedSensorVelocity();
    }

    /**
     * Get The Current Left Drive Velocity In Encoder Units
     * @return Encoder Units per 100ms
     */
    public double getLeftVelocity(){
        return driveLM.getSelectedSensorVelocity();
    }

    /**
     * Get The Right Drive RPM
     * @return Current Revolutions Per Minute
     */
    public double getRightRPM(){
        return driveRM.getSelectedSensorVelocity() / Constants.kFalconEncoderUPR * 600.0;
    }

    /**
     * Get The Left Drive RPM
     * @return Current Revolutions Per Minute
     */
    public double getLeftRPM(){
        return driveLM.getSelectedSensorVelocity() / Constants.kFalconEncoderUPR * 600.0;
    }

    /**
     * Reset DriveBase Sensors (true = reset, false = dont)
     * @param reset_encoders ?
     * @param reset_gyro ?
     */
    public void reset(boolean reset_encoders, boolean reset_gyro){
        if(reset_encoders){
            driveRM.setSelectedSensorPosition(0);
            driveLM.setSelectedSensorPosition(0);
        }
        if(reset_gyro){
            gyro.reset();
        }
    }

    /**
     * Get the NavX's Yaw Angle
     * @return Robot's Current Angle
     */
    public double getAngle(){
        return gyro.getYaw();
    }

    /**
     * Get the NavX's Heading
     * @return Accumulated Yaw Angle
     * Note: This Function can exeed values of 360
     */
    public double getHeading(){
        return gyro.getAngle();
    }

    /**
     * Get The Current Rate of change in the NavX's Yaw Axis
     * @return Robot's current turning rate
     */
    public double getAngleRate(){
        return gyro.getRate();
    }

    public SimpleMotorFeedforward getFeedForward(){
        return feedforward;
    }
    
    public PIDController getRightPID(){
        return rightPID;
    }

    public PIDController getLeftPID(){
        return leftPID;
    }

    public DifferentialDriveKinematics getKinematics(){
        return iKinematics;
    }

    /**
     * Returns Est Position of robot on feild.. in encoder units..?
     * @return
     */
    public Pose2d getGeoPos(){
        return iOdometry.getPoseMeters();
    }

    public void resetOdometry(Pose2d pose){
        reset(true, false);
        iOdometry.resetPosition(pose, Rotation2d.fromDegrees(getAngle()));
    }

    public DifferentialDriveWheelSpeeds getSpeeds(){
        return iSpeeds;
    }

    @Override
    public void periodic(){
        iOdometry.update(Rotation2d.fromDegrees(getAngle()), getLeftPosition(), getRightPosition());
    }
}

File: 6364 Worlds/src/main/cpp/subsystems/DriveSubsystem.cpp
/**
 * Drive Subsystem
 * 
 * @link DriveSubsystem.hpp
 * 
 * @author Jacob Chisholm
 * TEAM: 6364
 * 
 * Drive Subsystem:
 *    - 4 Falcon FX 500 Motors
 *    - Master and Slave configuration
 *    - Exponential Controller with Deadband
 *    - NavX Gyro
 *    
 */

#include "subsystems/DriveSubsystem.hpp"
#include "tools/Motors.hpp"
#include <frc/SPI.h>
#include <frc/smartdashboard/SmartDashboard.h>

DriveSubsystem::DriveSubsystem() :
driveRM{RobotMap::CAN::driveRM},
driveRS{RobotMap::CAN::driveRS},
driveLM{RobotMap::CAN::driveLM},
driveLS{RobotMap::CAN::driveLS},
navX{frc::SPI::kMXP} {

    motorConfiguration::Talon::rightDrive(driveRM);
    motorConfiguration::Talon::rightDrive(driveRS);
    motorConfiguration::Talon::leftDrive(driveLM);
    motorConfiguration::Talon::leftDrive(driveLS);

    driveRS.Follow(driveRM);
    driveLS.Follow(driveLM);

    //frc::SmartDashboard::PutBoolean("Reset Gyro Readings", false);
    SetName("DriveBase");
}

void DriveSubsystem::set(ControlMode mode, double leftPow, double rightPow){
    driveRM.Set(mode, rightPow);
    driveLM.Set(mode, leftPow);
}

void DriveSubsystem::set(double leftPow, double rightPow){
    driveRM.Set(ControlMode::Velocity, rightPow);
    driveLM.Set(ControlMode::Velocity, leftPow);
}

void DriveSubsystem::set(ControlMode mode, double iPow){
    driveRM.Set(mode, iPow);
    driveLM.Set(mode, iPow);
}

void DriveSubsystem::set(double velocity){
    driveRM.Set(ControlMode::Velocity, velocity);
    driveLM.Set(ControlMode::Velocity, velocity);
}

double DriveSubsystem::getRightVel(){
    return driveRM.GetSelectedSensorVelocity();
}

double DriveSubsystem::getLeftVel(){
    return driveLM.GetSelectedSensorVelocity();
}

double DriveSubsystem::getRightPos(){
    return driveRM.GetSelectedSensorPosition();
}

double DriveSubsystem::getLeftPos(){
    return driveLM.GetSelectedSensorPosition();
}

double DriveSubsystem::getPos(){
    return (driveRM.GetSelectedSensorPosition() + driveLM.GetSelectedSensorPosition())/2;
}

void DriveSubsystem::resetDrive(){
    driveRM.SetSelectedSensorPosition(0);
    driveLM.SetSelectedSensorPosition(0);
}

double DriveSubsystem::getRotation(){
    return navX.GetYaw();
}

void DriveSubsystem::resetNav(){
    navX.Reset();
}

void DriveSubsystem::calibrateNav(){
    navX.Calibrate();
}

float DriveSubsystem::getNavPos(Direction d){
    if(d == Direction::X){
        return navX.GetDisplacementX();
    }
    else if(d == Direction::Y){
        return navX.GetDisplacementY();
    }
    else if(d == Direction::Z){
        return navX.GetDisplacementZ();
    }
    else{
        return 0;
    }
}

void DriveSubsystem::resetNavPos(){
    navX.ResetDisplacement();
}

//float maxPitch = 0;
//float minPitch = 0;

void DriveSubsystem::Periodic(){
    frc::SmartDashboard::PutNumber("Gyro Tilt", navX.GetRoll());
    /*
    frc::SmartDashboard::PutNumber("Max Tilt", maxPitch);
    frc::SmartDashboard::PutNumber("Min Tilt", minPitch);
    if(navX.GetRoll() > maxPitch){
        maxPitch = navX.GetRoll();
    }
    if(navX.GetRoll() < minPitch){
        minPitch = navX.GetRoll();
    }

    if(frc::SmartDashboard::GetBoolean("Reset Gyro Readings", false) == true){
        maxPitch = 0;
        minPitch = 0;
         frc::SmartDashboard::PutBoolean("Reset Gyro Readings", false);
    }*/
}